Revision of metis_cal_chophome

oczoske opened the pull request:
Trying to make the recipe (renamed from `metis_img_chophome` to `metis_cal_chophome`, because it is not IMG specific) consistent with calibration plan and template manual. For instance, these foresee `WCU_OFF` frames, as for all other templates using the WCU, hence no need for `MASTER_DARK`. The product is currently defined as an image (a stack of all input images), with the primary result (the offsets in x and y) given as QC parameters in the header. This could be changed to a table.

hugobuddel commented:
I know this is still marked as draft, but I'll comment anyway, since the accompanying PRs are marked as ready, https://github.com/AstarVienna/METIS_Pipeline_Test_Data/pull/6 and https://github.com/AstarVienna/METIS_Simulations/pull/81 . The scripts in METIS_Simulations that check against the DRLD will fail if this PR is not merged as well.

However, the tests in this PR fail because it expects the labels to be on the same line as the sections. E.g.
```latex
\paragraph{\PROD{LM_SCI_BASIC_REDUCED}}\label{dataitem:lm_sci_basic_reduced}
```

instead of 
```
\subsubsection{Final product data item structures}
\label{sssec:imgfinaldatastructs}
```

We could update the tests to work also when the label is on a different line, but there is no reason to do that in this PR. Could you please change the labels back @oczoske ? (Or update the tests.)

hugobuddel commented on Recipes_Technical.tex:
```diff
@@ -47,50 +47,69 @@ \subsubsection{\REC*{metis_pupil_imaging}: Pupil imaging}\label{rec:metis_pupil_
                 & Calculate mirror offset\\
-Output data:	& Offset of the chopper mirror to be piped either into the \ac{ICS} for correction \\
-                & or to be used in the pipeline for astrometric correction\\
+Output data:	& Stacked chopperhome image \\
+                & Offset of the chopper mirror (as header keywords) to be piped either into the \ac{ICS} for correction  or to be used in the pipeline for astrometric correction\\
```
Perhaps this is the time to actually put something sensible here?

I have explicitly hardcoded an exception for this particular recipe, and I would like to get rid of it: https://github.com/AstarVienna/METIS_DRLD/blob/90b82bf96e1ecb04ee097be7c7a73919ea44613d/codes/drld_parser/hacks.py#L194-L197

Also, currently the tests fail because the hardcoded exception is for the old name of the recipe.

hugobuddel commented:
The new data items have not been properly propagated to the other tables:

```
python -m pytest -s -x --pdb codes/tests/test_datareductionlibra
rydesign.py::TestDataReductionLibraryDesign::test_dataitems_sanity
```

Shows:
```
BADPIX_MAP_det claims to be input for ['metis_det_lingain', 'metis_det_dark', 'metis_ifu_wavecal', 'metis_ifu_rsrf', 'metis_ifu_reduce', 'me
tis_ifu_distortion', 'metis_lm_img_flat', 'metis_lm_img_basic_reduce', 'metis_lm_img_distortion', 'metis_n_img_flat', 'metis_n_img_chopnod',
 'metis_n_img_distortion', 'metis_lm_lss_rsrf', 'metis_lm_lss_trace', 'metis_lm_lss_wave', 'metis_lm_lss_std', 'metis_lm_lss_sci', 'metis_n_
lss_rsrf', 'metis_n_lss_trace', 'metis_n_lss_std', 'metis_n_lss_sci', 'metis_lm_adc_slitloss', 'metis_n_adc_slitloss']
Input for:    & \REC{metis_det_lingain} \\
              & \REC{metis_det_dark} \\
              & \REC{metis_ifu_wavecal} \\
              & \REC{metis_ifu_rsrf} \\
              & \REC{metis_ifu_reduce} \\
              & \REC{metis_ifu_distortion} \\
              & \REC{metis_lm_img_flat} \\
              & \REC{metis_lm_img_basic_reduce} \\
              & \REC{metis_lm_img_distortion} \\
              & \REC{metis_n_img_flat} \\
              & \REC{metis_n_img_chopnod} \\
              & \REC{metis_n_img_distortion} \\
              & \REC{metis_LM_lss_rsrf} \\
              & \REC{metis_LM_lss_trace} \\
              & \REC{metis_LM_lss_wave} \\
              & \REC{metis_LM_lss_std} \\
              & \REC{metis_LM_lss_sci} \\
              & \REC{metis_N_lss_rsrf} \\
              & \REC{metis_N_lss_trace} \\
              & \REC{metis_N_lss_std} \\
              & \REC{metis_N_lss_sci} \\
              & \REC{metis_cal_chophome} \\
              & \REC{metis_lm_adc_slitloss} \\
              & \REC{metis_n_adc_slitloss} \\

LM_WCU_OFF_RAW claims to be input for ['metis_det_lingain', 'metis_lm_img_distortion', 'metis_lm_adc_slitloss']
Input for:    & \REC{metis_det_lingain} \\
              & \REC{metis_lm_img_distortion} \\
              & \REC{metis_cal_chophome} \\
              & \REC{metis_lm_adc_slitloss} \\

LM_CHOPHOME_COMBINED claims to be created by ['metis_cal_chophome']
```

It seems that `metis_cal_chophome` is not yet listed as a consumer of `LM_WCU_OFF_RAW` and `BADPIX_MAP_det` (I guess it should be `BADPIX_MAP_LM` there). You should be able to just copy those `Input for:` sections directly into the DRLD sections for those two data items.

Also, it seems that `LM_CHOPHOME_COMBINED` is not listed as output for `metis_cal_chophome`.

hugobuddel commented:
Since I already committed my typo-suggestion, I have also taken the liberty to update the tests directly in this branch. Hope that is okay.

The tests now work with the labels on the next line from the paragraph header. So it is not necessary to revert that change.

The `compare_yaml_with_drld.py` script of METIS_Simulations now passes in https://github.com/AstarVienna/METIS_Simulations/pull/81

These tests currently fail:
```
FAILED codes/tests/test_datareductionlibrarydesign.py::TestDataReductionLibraryDesign::test_dataitem_extraction - AssertionError: assert not ['LM_SCI_BASIC_REDUCED', 'N_SCI_BKG_SUBTRACTED', 'LM_SCI_OBJECT_CAT', 'LM_SCI_CALIBRATED', 'N_SCI_CALIBRATED...
FAILED codes/tests/test_datareductionlibrarydesign.py::TestDataReductionLibraryDesign::test_all_recipe_data_has_a_name - AssertionError: assert not [('metis_cal_chophome', DataItemReference(name=None, dtype='PROD', hyperref=None, description='Stacked chopp...
FAILED codes/tests/test_datareductionlibrarydesign.py::TestDataReductionLibraryDesign::test_dataitems_sanity - AssertionError: Found 6 problems with the dataitems internal consistency.
FAILED codes/tests/test_datareductionlibrarydesign.py::test_tikz - AttributeError: 'NoneType' object has no attribute 'replace'
```

The test is not marked as failing on github, because it is explicitly configured with `continue-on-error: true`, so you have to run them yourself or look at the output of github.

oczoske commented:
Sorry for the delay. I've added the inputs and outputs marked by that test to the respective tables and the tests now pass, with the exception of `test_datareductionlibrarydesign.py::TestDataReductionLibraryDesign::test_dataitem_extraction` (but that fails with many more items, not just the CHOPHOME products).   I have removed the exception in `hacks.py`.

hugobuddel commented:
Thanks! The test_dataitem_extraction test does pass on the master branch, so not sure what is going on.

Oh I think I know; I recall that test uses the paragraph headers to find dataitems, and you changed many of those. I'll have a look

oczoske commented:
Oh, that has to do with the labels being on the next line. `if "paragraph" in line and "dataitem" in line` where `dataitem` is taken from the label. I'll just put the label back on the same line.

hugobuddel commented:
I'm getting this error when running `pdflatex`:
```
! Misplaced \crcr.
\adl@LTlastrow ->\crcr                        
                       \noalign { \ifdim \adl@height =\z@ \ifdim \adl@depth ...              
l.1218 \end{datastructdef}                                                                   
```
I do not get that error on the master branch. Any idea what causes this?

hugobuddel commented:
> Oh, that has to do with the labels being on the next line. `if "paragraph" in line and "dataitem" in line` where `dataitem` is taken from the label. I'll just put the label back on the same line.

Yeah but it should not fail on that. So I fixed the tests in #318. That test is rather ugly, and should be redundant. As in, if that test fails, another (a bit better written) test should fail as well. So maybe we should throw it out at some point.

hugobuddel commented:
> > I'm getting this error when running `pdflatex`:
> > ```
> > ! Misplaced \crcr.
> > \adl@LTlastrow ->\crcr                        
> >                        \noalign { \ifdim \adl@height =\z@ \ifdim \adl@depth ...              
> > l.1218 \end{datastructdef}                                                                   
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I do not get that error on the master branch. Any idea what causes this?
> 
> Compiles here... Maybe because I've used `\lstinline` instead of the dreaded `\texttt`? Try now.

Yeah strange..

It does work on the CI: https://github.com/AstarVienna/METIS_DRLD/actions/runs/13183794721/job/36800907908 (requires #319 ...), but that uses tectonic IIRC.

It is indeed fine now. Not sure whether it is the `\lstinline` or the escaping of `_`. But anyway it works

oczoske commented:
`\lstinline` hasn't always played nice in tabular or other environments. Might be a version issue as well. Anyway, now it's consistent with all the other entries, which is also of value.

hugobuddel commented on IMG_data_items.tex:
```diff
@@ -882,319 +883,367 @@ \subsubsection{Master calibration / intermediate data item structures}\label{sss
+  OCA keywords: & \FITS{PRO.CATG}, \FITS{DRS.FILTER}\\
+  \FITS{DO.CATG}: & \CODE{N_SCI_CALIBRATED}\\[0.3cm]
+  % Template: & \TPL{METIS_img_n_cal_standard}\\
+  % & \TPL{METIS_img_n_*_obs_*}       \\
```
Found one case where your auto formatter made things worse! But overall it is for the better.

hugobuddel reviewed (approved):
Awesome, much better and much clearer. Thanks for fixing this @oczoske

hugobuddel commented:
By the way, there is a file https://github.com/AstarVienna/METIS_DRLD/blob/master/tikz/all.tex which contains all the tikz figures. It was useful in quickly iterating on just the figures. You could consider adding `metis_cal_chophome.tex` to that as well.
